dwave/optimization/src/nodes/manipulation.cpp

```diff
@@ -18,6 +18,154 @@
 
 namespace dwave::optimization {
 
+std::vector<ssize_t> make_concatenate_shape(std::vector<ArrayNode*> array_ptrs, ssize_t axis) {
```

Comment:
We should use `span` here. It ends up being more flexible by allowing other range types. It also avoids a copy. IMO it also does a better job of signalling to the user/reader what the intent is.
```c++
ConcatenateNode::ConcatenateNode(std::span<const ArrayNode*> array_ptrs, const ssize_t axis)
```

Reply:
fixed